{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from urllib.request import Request, urlopen\n",
    "from bs4 import BeautifulSoup\n",
    "import os\n",
    "import wget\n",
    "import subprocess\n",
    "import timeit\n",
    "import time\n",
    "\n",
    "'''\n",
    "traversal function:\n",
    "    iterate through Refseq online driectories, grabbing the correct links and current pages being displayed\n",
    "'''\n",
    "def traversal(taxa, spcies, ref_or_gen):\n",
    "    # Get all links (which are the taxonomic items)\n",
    "    input_search = spcies.replace(' ', '_')\n",
    "    \n",
    "    url = f\"https://ftp.ncbi.nlm.nih.gov/genomes/{ref_or_gen}/\" + taxa + '/' + input_search + '/latest_assembly_versions'\n",
    "\n",
    "    req = Request(url, headers={'User-Agent': 'Mozilla/5.0'})\n",
    "    html = urlopen(req).read()\n",
    "    soup = BeautifulSoup(html, 'lxml')\n",
    "\n",
    "    all_links = soup.find_all(\"a\")\n",
    "    \n",
    "    dir_of_interest = ''\n",
    "    for i in str(all_links[1])[9:]:\n",
    "        if i == '/':\n",
    "            break\n",
    "        dir_of_interest = dir_of_interest + i\n",
    "\n",
    "    dir_of_interest = dir_of_interest.replace(' ', '_')\n",
    "    # Get all links (which are the taxonomic items)\n",
    "    new_url = f\"https://ftp.ncbi.nlm.nih.gov/genomes/{ref_or_gen}/\" + taxa + '/' + input_search + '/latest_assembly_versions/' + dir_of_interest\n",
    "    req = Request(new_url, headers={'User-Agent': 'Mozilla/5.0'})\n",
    "    html = urlopen(req).read()\n",
    "    soup = BeautifulSoup(html, 'lxml')\n",
    "\n",
    "    further_links = soup.find_all(\"a\")\n",
    "    \n",
    "    url = new_url + '/' + dir_of_interest + '_protein.faa.gz'\n",
    "\n",
    "    file_n = dir_of_interest\n",
    "    file_faa = dir_of_interest + '_protein.faa'\n",
    "    # Check if file exists, and if not, download and/or unzip it using gzip\n",
    "    if os.path.exists('./absense_fasta_downloads/' + file_faa):\n",
    "        print(file_n + ' is already unzipped in your directory!')\n",
    "        return spcies, file_n\n",
    "    elif os.path.exists('./absense_fasta_downloads/' + dir_of_interest + '_protein.faa.gz'):\n",
    "        print('Already downloaded ' + dir_of_interest + '_protein.faa.gz. Unzipping now.')\n",
    "        zip_dir = dir_of_interest + '_protein.faa.gz'\n",
    "        !gunzip ./absense_fasta_downloads/$zip_dir\n",
    "        return spcies, file_n\n",
    "    else:\n",
    "        try:\n",
    "            wget.download(url, './absense_fasta_downloads')\n",
    "            print('Successfully downloaded ' + dir_of_interest + '_protein.faa.gz. Unzipping now.')\n",
    "            zip_dir = dir_of_interest + '_protein.faa.gz'\n",
    "            !gunzip ./absense_fasta_downloads/$zip_dir\n",
    "            print('Successfully grabbed annotated FASTA file called ' + zip_dir + ' for species ' + input_search)\n",
    "            return spcies, file_n\n",
    "        except:\n",
    "            return 0, 0\n",
    "       \n",
    "'''\n",
    "create_hash function:\n",
    "    loop through the online taxonomic groups and create a hash table for faster accession in future iterations\n",
    "    allows the program to correctly traverse the online directories based on only the inputted species name\n",
    "'''\n",
    "def create_hash(ref_or_gen):\n",
    "    # Iterate through online directories and construct a hash table (dictionary) from species to groupings\n",
    "    # We can ignore archaea, bacteria, and viruses because abSENSE does not work correctly with their genomes\n",
    "    taxa_groups = ['fungi', 'invertebrate', 'plant', 'protozoa', 'vertebrate_mammalian', 'vertebrate_other']\n",
    "\n",
    "    species_dict = {}\n",
    "    \n",
    "    if ref_or_gen == 'REF':\n",
    "        ins = 'Refseq'\n",
    "    else:\n",
    "        ins = 'Genbank'\n",
    "    \n",
    "    for k in taxa_groups:\n",
    "        # Get all links (which are the taxonomic items within each subgroup)\n",
    "        url = f\"https://ftp.ncbi.nlm.nih.gov/genomes/{ins.lower()}/{k}/\"\n",
    "        req = Request(url, headers={'User-Agent': 'Mozilla/5.0'})\n",
    "        html = urlopen(req).read()\n",
    "        soup = BeautifulSoup(html, 'lxml')\n",
    "        \n",
    "        all_links = soup.find_all(\"a\")\n",
    "        \n",
    "        # Loop through and create a mapping from each species to its group\n",
    "        for i in all_links:\n",
    "            val = str(i).strip('<a href=\"')[0:-5].split('/\">').pop().replace('_', ' ')\n",
    "            if 'annotation' not in val and 'Parent' not in val and 'summary' not in val:\n",
    "                species_dict[val] = k\n",
    "                \n",
    "    with open(f\"./absense_fasta_downloads/SPECIES_TAXA_HASH_{ref_or_gen}.txt\",\"w\") as f:\n",
    "        for i, j in species_dict.items():\n",
    "            f.write(i + ', ' + j + '\\n')\n",
    "        f.close()\n",
    "   \n",
    "'''\n",
    "open_hash function:\n",
    "    loop through the locally downloaded HASH text file and transform it to a dictionary\n",
    "'''\n",
    "def open_hash(inp_file):\n",
    "    # Iterate through SPECIES_TAXA_HASH.txt and create Python dictionary object\n",
    "    d = {}\n",
    "    with open(inp_file) as f:\n",
    "        for line in f:\n",
    "            key, value = line.strip().split(\",\")\n",
    "            d[key] = value.strip(' ')\n",
    "    return d\n",
    "\n",
    "'''\n",
    "get_link function:\n",
    "    from the large assembly dataframe, grab the relevant accession URLs to grab the protein FASTA files\n",
    "'''\n",
    "def get_link(data_assembly, input_search):\n",
    "\n",
    "    # Assure that the species with full genome representations are chosen, and get proper link\n",
    "    spec_data = data_assembly[data_assembly['organism_name'].fillna('na').str.contains(input_search)]\n",
    "    \n",
    "    # Add condition for if species is not in database\n",
    "    if len(spec_data) == 0:\n",
    "        return 0, 0\n",
    "    fin_data = spec_data[spec_data['genome_rep'] == 'Full']\n",
    "    \n",
    "    # Add condition for if genome is not complete\n",
    "    if len(fin_data) == 0:\n",
    "        return 0, 0\n",
    "\n",
    "    if sum(fin_data['refseq_category'] == 'reference genome') > 0:\n",
    "        dir_data = fin_data[fin_data['refseq_category'] == 'reference genome']\n",
    "    elif sum(fin_data['refseq_category'] == 'representative genome') > 0:\n",
    "        dir_data = fin_data[fin_data['refseq_category'] == 'representative genome']\n",
    "    elif sum(fin_data['relation_to_type_material'].isna()) != len(list(fin_data['relation_to_type_material'])):\n",
    "        dir_data = fin_data[fin_data['relation_to_type_material'].notna()]\n",
    "    else:\n",
    "        dtes = list(pd.to_datetime(list(fin_data['seq_rel_date'])))\n",
    "        dir_data = fin_data[pd.to_datetime(fin_data['seq_rel_date']) == max(dtes)]\n",
    "    \n",
    "    ftp = dir_data['ftp_path'].iloc[0]\n",
    "    \n",
    "    if 'vir' in dir_data['organism_name'].iloc[0]:\n",
    "        return 0, 0\n",
    "    \n",
    "    # Construct link for protein annotation\n",
    "    acc_name = dir_data['# assembly_accession'].iloc[0] + '_' + dir_data['asm_name'].iloc[0]\n",
    "    lnk = 'https' + ftp[3:] + '/' + acc_name.replace(' ', '_') + '_protein.faa.gz'\n",
    "    return lnk, acc_name\n",
    "\n",
    "'''\n",
    "alt_search function:\n",
    "    if species protein annotation was not found in the assembly dataframe, manually traverse the online directories\n",
    "    in search of the recent assembly download\n",
    "'''\n",
    "def alt_search(input_search, ref_or_gen):\n",
    "    # Check if Refseq or Genbank option\n",
    "    if ref_or_gen == 'REF':\n",
    "        ins = 'Refseq'\n",
    "    else:\n",
    "        ins = 'Genbank'\n",
    "    \n",
    "    if os.path.exists(f'./absense_fasta_downloads/SPECIES_TAXA_HASH_{ref_or_gen}.txt'):\n",
    "        print(f'{ins} species hash mappings already created, traversing online directories now...')\n",
    "    else:\n",
    "        print(f'Creating species taxa hash mappings for {ins}...')\n",
    "        create_hash(ref_or_gen)\n",
    "      \n",
    "    # Iterate through hash table and use it for a file traversal to check if annotation exists\n",
    "    hash_tbl = open_hash(f\"./absense_fasta_downloads/SPECIES_TAXA_HASH_{ref_or_gen}.txt\")\n",
    "    try:\n",
    "        taxa_cat = hash_tbl[input_search]\n",
    "        res_spc, res_acc = traversal(taxa_cat, input_search, ins.lower())\n",
    "        if res_acc != 0:\n",
    "            return res_spc, res_acc\n",
    "        else:\n",
    "            print(f'Alternative {ins} database route failed.')\n",
    "            return 0, 0\n",
    "    except:\n",
    "        print(f'Invalid species name, cannot be found in {ins} alternate directory.')\n",
    "        return 0, 0\n",
    "    \n",
    "'''\n",
    "get_annotated_seq function:\n",
    "    main function that searchs Genbank and Refseq assembly summary files to grab\n",
    "    relevant accession information for an inputted species, and attempts alternative searches\n",
    "    if no recent downloads are available from the assembly file\n",
    "'''\n",
    "def get_annotated_seq(input_search):\n",
    "    ##################\n",
    "    ### Refseq search\n",
    "    \n",
    "    input_search = input_search.replace('_', ' ')\n",
    "    \n",
    "    print('--------------------------')\n",
    "    print(f'First searching Refseq for {input_search}...')\n",
    "    \n",
    "    if not os.path.exists('absense_fasta_downloads'):\n",
    "        !mkdir ./absense_fasta_downloads/\n",
    "    \n",
    "    # Download the file if it does not exist\n",
    "    if os.path.exists('./absense_fasta_downloads/assembly_summary_refseq.txt'):\n",
    "        print('Refseq assembly summary already downloaded! Grabbing species now.')\n",
    "    else:\n",
    "        print('Grabbing Refseq assembly from online...')\n",
    "        # Download file\n",
    "        try:\n",
    "            wget.download(\"https://ftp.ncbi.nlm.nih.gov/genomes/refseq/assembly_summary_refseq.txt\", './absense_fasta_downloads/')\n",
    "        except:\n",
    "            print('Error downloading assembly_summary_refseq.txt from https://ftp.ncbi.nlm.nih.gov/genomes/refseq/. Please try manually downloading into the absense_fasta_downloads directory if the problem persists.')\n",
    "    data_assembly = pd.read_csv('./absense_fasta_downloads/assembly_summary_refseq.txt', sep='\\t', skiprows=1)\n",
    "    \n",
    "    under_input_search = input_search.replace(' ', '_')\n",
    "    \n",
    "    # Grab link\n",
    "    lnk, acc_name = get_link(data_assembly, input_search)\n",
    "    \n",
    "    # Ensure no spaces\n",
    "    if acc_name != 0:\n",
    "        acc_name = acc_name.replace(' ', '_')\n",
    "    \n",
    "    if lnk != 0 and acc_name != 0:\n",
    "        # Check if annotation is present\n",
    "        if os.path.exists('./absense_fasta_downloads/' + acc_name + '_protein.faa'):\n",
    "            print('Refseq file called ' + acc_name + '_protein.faa already in your directory!')\n",
    "            return input_search, acc_name\n",
    "        elif os.path.exists('./absense_fasta_downloads/' + acc_name + '_protein.faa.gz'):\n",
    "            print('Zipped Refseq file called ' + acc_name + '_protein.faa already in your directory! Unzipping now.')\n",
    "            zip_dir = acc_name + '_protein.faa.gz'\n",
    "            !gunzip ./absense_fasta_downloads/$zip_dir\n",
    "            return input_search, acc_name\n",
    "        else:\n",
    "            try:\n",
    "                wget.download(lnk, './absense_fasta_downloads/')\n",
    "                zip_dir = acc_name + '_protein.faa.gz'\n",
    "                !gunzip ./absense_fasta_downloads/$zip_dir\n",
    "                print('Successfully downloaded and unzipped Refseq file called ' + acc_name + '_protein.faa (for species ' + input_search + ')')\n",
    "                return input_search, acc_name\n",
    "            except:\n",
    "                print('No Refseq annotation exists in the directory for ' + acc_name + ' (' + input_search + ')')\n",
    "                print('Now searching Genbank...')\n",
    "    else:\n",
    "        print('No Refseq annotation exists for ' + input_search)\n",
    "        print(f'Now searching Genbank for {input_search}...')\n",
    "\n",
    "    ##################\n",
    "    ### Genbank search\n",
    "    \n",
    "    # Download the file if it does not exist\n",
    "    if os.path.exists('./absense_fasta_downloads/assembly_summary_genbank.txt'):\n",
    "        print('Genbank assembly summary already downloaded! Grabbing species now.')\n",
    "    else:\n",
    "        print('Grabbing Genbank assembly from online...')\n",
    "        # Download file\n",
    "        try:\n",
    "            wget.download(\"https://ftp.ncbi.nlm.nih.gov/genomes/genbank/assembly_summary_genbank.txt\", './absense_fasta_downloads/')\n",
    "        except:\n",
    "            print('Error downloading assembly_summary_refseq.txt from https://ftp.ncbi.nlm.nih.gov/genomes/genbank/. Please try manually downloading into the absense_fasta_downloads directory if the problem persists.')\n",
    "    data_assembly = pd.read_csv('./absense_fasta_downloads/assembly_summary_genbank.txt', sep='\\t', skiprows=1)\n",
    "    \n",
    "    # Grab link\n",
    "    lnk, acc_name = get_link(data_assembly, input_search)\n",
    "    \n",
    "    # Ensure no spaces\n",
    "    if acc_name != 0:\n",
    "        acc_name = acc_name.replace(' ', '_')\n",
    "    \n",
    "    if lnk != 0 and acc_name != 0:\n",
    "        # Check if annotation is present\n",
    "        if os.path.exists('./absense_fasta_downloads/' + acc_name + '_protein.faa'):\n",
    "            print('Genbank file called ' + acc_name + '_protein.faa already in your directory!')\n",
    "            return input_search, acc_name\n",
    "        elif os.path.exists('./absense_fasta_downloads/' + acc_name + '_protein.faa.gz'):\n",
    "            print('Zipped Genbank file called ' + acc_name + '_protein.faa already in your directory! Unzipping now.')\n",
    "            zip_dir = acc_name + '_protein.faa.gz'\n",
    "            !gunzip ./absense_fasta_downloads/$zip_dir\n",
    "            return input_search, acc_name\n",
    "        else:\n",
    "            try:\n",
    "                wget.download(lnk, './absense_fasta_downloads/')\n",
    "                zip_dir = acc_name + '_protein.faa.gz'\n",
    "                !gunzip ./absense_fasta_downloads/$zip_dir\n",
    "                print('Successfully downloaded and unzipped Genbank file called ' + acc_name + '_protein.faa (' + input_search + ')')\n",
    "                return input_search, acc_name\n",
    "            except:\n",
    "                print('No Genbank annotation exists in the directory for ' + acc_name + ' (' + input_search + ')')\n",
    "    \n",
    "    ##################\n",
    "    ### Alternate latest_assembly_version search\n",
    "    \n",
    "    # Last ditch effort if there are no species in Refseq or Genbank through the standard way\n",
    "    # We essentially traverse a different directory and check latest_assembly_versions manually\n",
    "    print('Trying alternative database route for Refseq and/or Genbank...')\n",
    "    \n",
    "    spcies, acc_name = alt_search(input_search, 'REF')\n",
    "    if acc_name != 0:\n",
    "        return input_search, acc_name\n",
    "    \n",
    "    spcies, acc_name = alt_search(input_search, 'GEN')\n",
    "    if acc_name != 0:\n",
    "        return input_search, acc_name\n",
    "    \n",
    "    print('Invalid species name or species does not have a protein annotation in either Genbank or Refseq.')\n",
    "    return 0, 0\n",
    "\n",
    "'''\n",
    "get_eukary_odb10 function:\n",
    "    ensure BUSCO eukaryota genes are locally downloaded for ortholog detection\n",
    "'''\n",
    "def get_eukary_odb10():\n",
    "    # Download the file if it does not exist\n",
    "    if os.path.exists('eukaryota_odb10.2020-09-10.tar.gz'):\n",
    "        print('Zipped Eukaryota ODB10 BUSCO genes already downloaded! Unzipping now.')\n",
    "        !gunzip ./eukaryota_odb10.2020-09-10.tar.gz\n",
    "        !tar -xvf eukaryota_odb10.2020-09-10.tar\n",
    "        return True\n",
    "    elif os.path.exists('./eukaryota_odb10'):\n",
    "        print('Eukaryota ODB10 BUSCO genes already downloaded and successfully opened.')\n",
    "        return True\n",
    "    elif os.path.exists('eukaryota_odb10.2020-09-10.tar'):\n",
    "        print('Eukaryota ODB10 BUSCO genes already downloaded!')\n",
    "        !tar -xvf eukaryota_odb10.2020-09-10.tar\n",
    "        print('Successfully opened tar file.')\n",
    "        return True\n",
    "    else:\n",
    "        print('Grabbing BUSCO genes from its online directory (https://busco-data.ezlab.org/v4/data/lineages/)...')\n",
    "        \n",
    "        # Download file using wget\n",
    "        try:\n",
    "            print('Downloading...')\n",
    "            wget.download('https://busco-data.ezlab.org/v4/data/lineages/eukaryota_odb10.2020-09-10.tar.gz', './')\n",
    "            !gunzip ./eukaryota_odb10.2020-09-10.tar.gz\n",
    "            print('Successfully downloaded.')\n",
    "            !tar -xvf eukaryota_odb10.2020-09-10.tar\n",
    "            print('Successfully opened tar file.')\n",
    "            return True\n",
    "        except:\n",
    "            print('Error. File download failed from https://busco-data.ezlab.org/v4/data/lineages/.')\n",
    "            print('Eukaryotic file name may have changed. Should be eukaryota_odb10.2020-09-10.tar.gz.')\n",
    "            return False\n",
    " \n",
    "'''\n",
    "run_hmmer_busco function:\n",
    "    perform a HMMER homology search with each species protein annotation FASTA file with each\n",
    "    eukaryota BUSCO gene file and save results in a directory for future analysis\n",
    "'''\n",
    "def run_hmmer_busco(species_fasta_list):\n",
    "    # Create file\n",
    "    if not os.path.exists('absense_hmmer_out'):\n",
    "        !mkdir ./absense_hmmer_out/\n",
    "\n",
    "    # Iterate through species\n",
    "    for i in species_fasta_list:\n",
    "        # Go through each .hmm file\n",
    "        for k in os.listdir('./eukaryota_odb10/hmms'):\n",
    "            # Ensure proper directory traversal\n",
    "            dir_name = 'out_hmmer_' + i[0:-4]\n",
    "            file_name = 'res_hmmer_' + k.strip('.hmm') + '.txt'\n",
    "            if not os.path.exists(f'./absense_hmmer_out/{dir_name}'):\n",
    "                !mkdir ./absense_hmmer_out/$dir_name\n",
    "                \n",
    "            # Perform hmmsearch\n",
    "            !hmmsearch ./eukaryota_odb10/hmms/$k ./absense_fasta_downloads/$i > ./absense_hmmer_out/$dir_name/$file_name\n",
    "            \n",
    "'''\n",
    "traverse_hmmer_file function:\n",
    "    iterate through HMMER functions and grab genes in which match met the inputted E-value threshold\n",
    "    and also in which there is only 1 mutual hit\n",
    "'''\n",
    "def traverse_hmmer_file(file_name, e_val):\n",
    "    # Iterate through file, check E-val and quantity conditions, and return\n",
    "    # gene label and query file name\n",
    "    with open(file_name, \"r\") as file:\n",
    "        lines = []\n",
    "        for line in file:\n",
    "            lines.append(line)\n",
    "            \n",
    "        # Grab E-value and query name\n",
    "        if (lines[15] == '\\n') and (float(lines[14].split()[0]) < e_val):\n",
    "            return lines[14].split()[8], lines[5].split()[4][23:][0:-4]\n",
    "        else:\n",
    "            return 0, 0\n",
    "   \n",
    "'''\n",
    "scan_files function:\n",
    "    go through HMMER functions and find genes that are consistently hit across all species of interest,\n",
    "    acting as a heuristic method for ortholog detection with a query of genes of interest to search\n",
    "'''\n",
    "def scan_files(species, check_query, e_val, comb):\n",
    "    # Traverse the relevant files\n",
    "    dir_name = './absense_hmmer_out/out_hmmer_' + species[0:-4]\n",
    "    \n",
    "    gene = []\n",
    "    query = []\n",
    "    \n",
    "    # If check_query is empty, that means it is the cntrl species\n",
    "    if len(check_query) == 0:\n",
    "        for f in os.listdir(dir_name):\n",
    "            g, q = traverse_hmmer_file(dir_name + '/' + f, e_val)\n",
    "            if g != 0:\n",
    "                gene.append(g)\n",
    "            if q != 0:\n",
    "                query.append(q)\n",
    "            if g != 0 and q != 0:\n",
    "                comb[q] = []\n",
    "                comb[q].append(g)\n",
    "        return gene, query, comb\n",
    "    # If check_query is nonempty, then we want to compare only query files that\n",
    "    # successfully identified an ortholog\n",
    "    else:\n",
    "        for f in check_query:\n",
    "            g, q = traverse_hmmer_file(dir_name + '/' + 'res_hmmer_' + f + '.txt', e_val)\n",
    "            if g != 0:\n",
    "                gene.append(g)\n",
    "            if q != 0:\n",
    "                query.append(q)\n",
    "            if g != 0 and q != 0:\n",
    "                comb[q].append(g)\n",
    "        return gene, query, comb\n",
    "    \n",
    "'''\n",
    "grab_orthologs function:\n",
    "    starting with the source species, find all genes that were hit of the eukaryotic genes, and\n",
    "    repeat the process for all subsequent target species, only search genes that are orthologs of all\n",
    "    the species searched prior\n",
    "'''\n",
    "def grab_orthologs(all_species, e_val):\n",
    "    # Iterate over species to determine ortholog genes\n",
    "    \n",
    "    species_genes = ['']\n",
    "    query_genes = []\n",
    "    comb = {}\n",
    "    \n",
    "    cntrl_species = all_species[0]\n",
    "    other_species = all_species[1:]\n",
    "    \n",
    "    # Traverse the relevant files\n",
    "    species_genes[0], query, comb = scan_files(cntrl_species, [], e_val, comb)\n",
    "    query_genes.append(set(query))\n",
    "    \n",
    "    if len(other_species) > 0:\n",
    "        # Go over other species and widdle down relevant ortholog genes\n",
    "        for k in other_species:\n",
    "            tmp, query, comb = scan_files(k, query, e_val, comb)\n",
    "            \n",
    "            species_genes.append(tmp)\n",
    "            query_genes.append(set(query))\n",
    "\n",
    "    # Take intersection of the set to get only genes that are consistent across species\n",
    "    out_query = set.intersection(*query_genes)\n",
    "\n",
    "    return_vals = []\n",
    "    for i in out_query:\n",
    "        return_vals.append(comb[i])\n",
    "        \n",
    "    # Return an array, where each index is a list of the ortholog genes\n",
    "    return return_vals\n",
    "\n",
    "'''\n",
    "create_muscle_files function:\n",
    "    iterate through protein annotation files and grab all ortholog genes,\n",
    "    grouping them into their own FASTA files\n",
    "'''\n",
    "def create_muscle_files(adj_list, fasta_files, spcies_list):\n",
    "    transposed_list = np.array(adj_list).T\n",
    "    \n",
    "    # Create directory\n",
    "    dir_nm = 'absense_files_for_muscle_' + fasta_files[0][0:-4]\n",
    "    !rm -rf $dir_nm\n",
    "    if not os.path.exists(dir_nm):\n",
    "        !mkdir ./$dir_nm\n",
    "\n",
    "    # Iterate through ortholog genes \n",
    "    orthos = [['']*len(transposed_list)]*len(adj_list)\n",
    "    for i in range(len(transposed_list)):\n",
    "        with open('./absense_fasta_downloads/' + fasta_files[i], 'r') as file:\n",
    "            gene_grab = [False]*len(adj_list)\n",
    "            # Go through and reformat so we can write to proper files\n",
    "            for j in range(len(adj_list)):\n",
    "                for line in file:\n",
    "                    if transposed_list[i][j] in line:\n",
    "                        gene_grab[j] = True\n",
    "                    elif '>' in line:\n",
    "                        gene_grab[j] = False\n",
    "                    if gene_grab[j] == True and '>' not in line:\n",
    "                        orthos[j][i] = orthos[j][i] + line.strip('\\n').strip('*')\n",
    "\n",
    "    # Combine the ortholog genes into individual files\n",
    "    for k in range(len(transposed_list[0])):\n",
    "        f_name = f'./{dir_nm}/{transposed_list[0][k]}.faa'\n",
    "        file = open(f_name,'w')\n",
    "        file.close()\n",
    "        \n",
    "        with open(f_name, 'w') as f:\n",
    "            len_of_line = 70\n",
    "            for m in range(len(transposed_list)):\n",
    "                f.write('>' + spcies_list[m][0] + '_' + spcies_list[m].replace(' ', '_').split('_')[1][0:5] + adj_list[k][m] + '\\n')\n",
    "\n",
    "                # Write with proper line breaks\n",
    "                for p in range(0, len(orthos[k][m]), len_of_line):\n",
    "                    f.write(orthos[k][m][p:(p+len_of_line)])\n",
    "                    \n",
    "                    if p == len(orthos[k][m]) - (len(orthos[k][m]) % 70):\n",
    "                        f.write('\\n')\n",
    "\n",
    "'''\n",
    "compute_muscle function:\n",
    "    going through the created FASTA files for MUSCLE, perform MUSCLE locally on each\n",
    "    set of ortholog genes and store the multiple sequence alignment results\n",
    "'''\n",
    "def compute_muscle(accessions):\n",
    "    source_accession = accessions[0]\n",
    "    \n",
    "    # Directory in which the data resides\n",
    "    dirloc = f\"./absense_files_for_muscle_{source_accession[0:-4]}\" \n",
    "    fl_str = f\"absense_files_for_muscle_{source_accession[0:-4]}\"\n",
    "\n",
    "    count = 0\n",
    "    lst_file = []\n",
    "    # Check if valid files\n",
    "    for file in os.scandir(dirloc):\n",
    "        if (file.path.endswith(\".faa\")) and file.is_file():\n",
    "            count = count + 1\n",
    "            lst_file.append(file.name)\n",
    "     \n",
    "    # Ensure correct directory exists\n",
    "    fl_name = './muscle_out_for_absense_' + fl_str[25:]\n",
    "    if not os.path.exists(fl_name):\n",
    "        !mkdir ./$fl_name\n",
    "\n",
    "    # Assumes the current directory the script resides in contains the 'muscle3.8.31_i86win32.exe' file\n",
    "    for i in range(count):\n",
    "        # Local version\n",
    "        in_dir = fl_str\n",
    "        out_dir = fl_name\n",
    "        val = lst_file[i]\n",
    "        out_nm = 'out_' + lst_file[i][0:9] + '.faa'\n",
    "        in_file = f'{in_dir}/{val}'\n",
    "        out_file = f'{out_dir}/{out_nm}.faa'\n",
    "\n",
    "        !muscle -in ./$in_dir/$val -out ./$out_dir/$out_nm\n",
    "        \n",
    "        # API Version (for possible future reference)\n",
    "        # os.system('cmd /k \"python muscle.py --email {email} {val}\"'.format(val = lst_file[i-1]))\n",
    "    \n",
    "'''\n",
    "compute_protdist function:\n",
    "    from the multiple sequence alignment files, run Protdist from PHYLIP\n",
    "    on the MSA files to get the evolutionary distances\n",
    "'''\n",
    "def compute_protdist(source_accession):\n",
    "    # Directory in which the data resides\n",
    "    dirloc = f\"./muscle_out_for_absense_{source_accession[0:-4]}\" \n",
    "    fl_str = f'muscle_out_for_absense_{source_accession[0:-4]}'\n",
    "\n",
    "    lst_file = []\n",
    "\n",
    "    for file in os.scandir(dirloc):\n",
    "        if (file.path.endswith(\".faa\")) and file.is_file():\n",
    "            lst_file.append(file)\n",
    "\n",
    "    # Init dictionary for gene names and respective sequences\n",
    "    genes = {}\n",
    "\n",
    "    # Read first file to get the gene names\n",
    "    with open(lst_file[0], 'r') as t_file:\n",
    "        for l in t_file:\n",
    "            if l[0] == '>':\n",
    "                genes[l[1:8]] = ''\n",
    "\n",
    "    # Iterate through the file contents\n",
    "    for i in range(len(lst_file)):\n",
    "        # Open file to read it\n",
    "        gene_track = ''\n",
    "        with open(lst_file[i], 'r') as r_file:\n",
    "            for line in r_file:\n",
    "                if line[0] == '>':\n",
    "                    gene_track = line[1:8]\n",
    "                else:\n",
    "                    genes[gene_track] = genes[gene_track] + line.strip('\\n').strip('*')\n",
    "\n",
    "    fl_name = './protdist_infile_outfile_' + fl_str[23:]\n",
    "    if not os.path.exists(fl_name):\n",
    "        !mkdir ./$fl_name\n",
    "\n",
    "    if os.path.exists('outfile'):\n",
    "        !rm outfile\n",
    "\n",
    "    # Write and format file for PHYLIP and Protdist\n",
    "    # Must have number of species and length of each sequence at the top left, followed by each name followed by a sequence\n",
    "    # *Has* to be named infile (as per Protdist documentation, seems like there's no way to change it)\n",
    "    with open('./infile', 'w') as w_file:\n",
    "        w_file.write('   ' + str(len(genes.keys())) + '   ' + str(len(genes[list(genes.keys())[0]])) + '\\n')\n",
    "        for k in genes:\n",
    "            w_file.write(k + '   ' + genes[k] + '\\n')\n",
    "\n",
    "    !echo Y | ./protdist\n",
    "\n",
    "    !cp infile infile ./$fl_name\n",
    "    !cp outfile outfile ./$fl_name\n",
    "    \n",
    "    return fl_name\n",
    "\n",
    "'''\n",
    "read_accession_file function:\n",
    "    read the accession file of all the species of interest and their corresponding \n",
    "    protein annotation FASTA file names for use in the pipeline\n",
    "'''\n",
    "def read_accession_file(file_name):\n",
    "    # Go through accession file and pass back the accession codes and species names\n",
    "    species_names = []\n",
    "    species_accessions = []\n",
    "    \n",
    "    # Iterate file and return valid accessions and species names\n",
    "    with open(file_name, 'r') as f:\n",
    "        k = 0\n",
    "        for line in f:\n",
    "            if k != 0:\n",
    "                # Split comma delineated file\n",
    "                txt = line.strip('\\n').strip(' ').split(',')\n",
    "                if txt[1] != '0_protein.faa':\n",
    "                    species_names.append(txt[0])\n",
    "                    species_accessions.append(txt[1])\n",
    "            k += 1\n",
    "    \n",
    "    return species_names, species_accessions\n",
    "\n",
    "'''\n",
    "grab_assemblies function:\n",
    "    iterate through species of interest and perform an online depth search through Refseq and Genbank\n",
    "    to download the protein annotations to a local directory\n",
    "'''\n",
    "def grab_assemblies(list_of_species, override_dict = {}):\n",
    "    # List of species contains list of species, where first species is the source and the others are the target\n",
    "    # override_dict is a dictionary of species and their corresponding FASTA files, for when you manually input\n",
    "    # FASTA files into the directory so the pipeline can incorporate them correctly\n",
    "    \n",
    "    species_names = [0]*len(list_of_species)\n",
    "    species_accessions = [0]*len(list_of_species)\n",
    "    \n",
    "    # Grab protein annotations for all species\n",
    "    for i in range(len(list_of_species)):\n",
    "        species_names[i], species_accessions[i] = get_annotated_seq(list_of_species[i])\n",
    "        \n",
    "    # Write output accession codes to a text file\n",
    "    with open('absense_accession_codes_infile.txt', 'w') as f:\n",
    "        f.write('name,accession,type\\n')\n",
    "        for k in range(len(species_names)):\n",
    "            if list_of_species[k] in override_dict:\n",
    "                f.write(f'{list_of_species[k]},{override_dict[list_of_species[k]][0]},{override_dict[list_of_species[k]][1]}\\n')\n",
    "            elif k != 0:\n",
    "                f.write(f'{list_of_species[k]},{species_accessions[k]}_protein.faa,target\\n')\n",
    "            else:\n",
    "                f.write(f'{list_of_species[k]},{species_accessions[k]}_protein.faa,source\\n')\n",
    "            \n",
    "    f.close()\n",
    "    \n",
    "    if 0 in species_accessions:\n",
    "        print('--------------------')\n",
    "        print('Warning: not all species had their protein annotations downloaded.')\n",
    "        print('Try manually downloading the necessary files and adding the species names and protein annotation file names to')\n",
    "        print('the absense_accession_codes_infile.txt file (or something else if it was changed).')\n",
    "        print('When satisfied, proceed with the evolutionary distance computation.')\n",
    "    else:\n",
    "        print('--------------------')\n",
    "        print('Successfully downloaded protein annotations for all species!')\n",
    "     \n",
    "'''\n",
    "run_evo_dist_computation function:\n",
    "    perform the pipeline of getting the eukaryotic BUSCO genes, performing HMMER homology search\n",
    "    to detect orthologs, getting the MSA of the orthologous genes, and then calculating the distances\n",
    "    based on the alignment\n",
    "'''\n",
    "def run_evo_dist_computation(evalue = 0.01, acc_file = 'absense_accession_codes_infile.txt'):\n",
    "    # Ensure e-value is below HMMER threshold\n",
    "    if evalue > 0.01:\n",
    "        print('Warning. E-value is above HMMER threshold of 0.01. Please reduce for meaningful results.')\n",
    "        return\n",
    "    \n",
    "    # Read accession codes file\n",
    "    print('-------------------')\n",
    "    species_names, species_accessions = read_accession_file(acc_file)\n",
    "        \n",
    "    # Grab BUSCO genes\n",
    "    success_busco = get_eukary_odb10()\n",
    "    \n",
    "    if success_busco == False:\n",
    "        print('Error downloading BUSCO genes from https://busco-data.ezlab.org/v4/data/lineages/')\n",
    "        print('Please try manually downloading into the local directory if the issue persists.')\n",
    "        print('Program cannot continue without the BUSCO eukaryota odb10 genes.')\n",
    "        return False\n",
    "    \n",
    "    print('-------------------')\n",
    "    print('Running HMMER calcuations now...')\n",
    "    run_hmmer_busco(species_accessions)\n",
    "    \n",
    "    print('-------------------')\n",
    "    print(f'Iterating results and grabbing orthologs now with E-value of {evalue}...')\n",
    "    result = grab_orthologs(species_accessions, evalue)\n",
    "    \n",
    "    print('-------------------')\n",
    "    print('Creating files for MUSCLE now...')\n",
    "    create_muscle_files(result, species_accessions, species_names)\n",
    "    \n",
    "    print('-------------------')\n",
    "    print('Running MUSCLE now...')\n",
    "    compute_muscle(species_accessions)\n",
    "    \n",
    "    print('-------------------')\n",
    "    print('Running Protdist now...')\n",
    "    out = compute_protdist(species_accessions[0])\n",
    "    \n",
    "    print('-------------------')\n",
    "    print(f'Done! Results are in \"outfile\" in the {out} directory. Thanks!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--------------------------\n",
      "First searching Refseq for Saccharomyces cerevisiae...\n",
      "Refseq assembly summary already downloaded! Grabbing species now.\n",
      "Refseq file called GCF_000146045.2_R64_protein.faa already in your directory!\n",
      "--------------------------\n",
      "First searching Refseq for Saccharomyces mikatae...\n",
      "Refseq assembly summary already downloaded! Grabbing species now.\n",
      "No Refseq annotation exists for Saccharomyces mikatae\n",
      "Now searching Genbank for Saccharomyces mikatae...\n",
      "Genbank assembly summary already downloaded! Grabbing species now.\n",
      "No Genbank annotation exists in the directory for GCA_000166975.1_ASM16697v1 (Saccharomyces mikatae)\n",
      "Trying alternative database route for Refseq and/or Genbank...\n",
      "Refseq species hash mappings already created, traversing online directories now...\n",
      "Invalid species name, cannot be found in Refseq alternate directory.\n",
      "Genbank species hash mappings already created, traversing online directories now...\n",
      "Alternative Genbank database route failed.\n",
      "Invalid species name or species does not have a protein annotation in either Genbank or Refseq.\n",
      "--------------------------\n",
      "First searching Refseq for Naumovozyma castellii...\n",
      "Refseq assembly summary already downloaded! Grabbing species now.\n",
      "Refseq file called GCF_000237345.1_ASM23734v1_protein.faa already in your directory!\n",
      "--------------------------\n",
      "First searching Refseq for Kluyveromyces lactis...\n",
      "Refseq assembly summary already downloaded! Grabbing species now.\n",
      "Refseq file called GCF_000002515.2_ASM251v1_protein.faa already in your directory!\n",
      "--------------------------\n",
      "First searching Refseq for Aspergillus nidulans...\n",
      "Refseq assembly summary already downloaded! Grabbing species now.\n",
      "Refseq file called GCF_000149205.2_ASM14920v2_protein.faa already in your directory!\n",
      "--------------------------\n",
      "First searching Refseq for Saccharomyces kudriavzevii...\n",
      "Refseq assembly summary already downloaded! Grabbing species now.\n",
      "No Refseq annotation exists for Saccharomyces kudriavzevii\n",
      "Now searching Genbank for Saccharomyces kudriavzevii...\n",
      "Genbank assembly summary already downloaded! Grabbing species now.\n",
      "No Genbank annotation exists in the directory for GCA_003327635.1_ASM332763v1 (Saccharomyces kudriavzevii)\n",
      "Trying alternative database route for Refseq and/or Genbank...\n",
      "Refseq species hash mappings already created, traversing online directories now...\n",
      "Invalid species name, cannot be found in Refseq alternate directory.\n",
      "Genbank species hash mappings already created, traversing online directories now...\n",
      "GCA_000167075.2_Saccharomyces_kudriavzevii_strain_IFO1802_v1.0 is already unzipped in your directory!\n",
      "--------------------------\n",
      "First searching Refseq for Yarrowia lipolytica...\n",
      "Refseq assembly summary already downloaded! Grabbing species now.\n",
      "Refseq file called GCF_000002525.2_ASM252v1_protein.faa already in your directory!\n",
      "--------------------------\n",
      "First searching Refseq for Schizosaccharomyces pombe...\n",
      "Refseq assembly summary already downloaded! Grabbing species now.\n",
      "Refseq file called GCF_000002945.1_ASM294v2_protein.faa already in your directory!\n",
      "--------------------------\n",
      "First searching Refseq for Eremothecium gossypii...\n",
      "Refseq assembly summary already downloaded! Grabbing species now.\n",
      "Refseq file called GCF_000091025.4_ASM9102v4_protein.faa already in your directory!\n",
      "--------------------------\n",
      "First searching Refseq for Saccharomyces paradoxus...\n",
      "Refseq assembly summary already downloaded! Grabbing species now.\n",
      "Refseq file called GCF_002079055.1_ASM207905v1_protein.faa already in your directory!\n",
      "--------------------------\n",
      "First searching Refseq for Saccharomyces bayanus...\n",
      "Refseq assembly summary already downloaded! Grabbing species now.\n",
      "No Refseq annotation exists for Saccharomyces bayanus\n",
      "Now searching Genbank for Saccharomyces bayanus...\n",
      "Genbank assembly summary already downloaded! Grabbing species now.\n",
      "No Genbank annotation exists in the directory for GCA_019431265.1_ASM1943126v1 (Saccharomyces bayanus)\n",
      "Trying alternative database route for Refseq and/or Genbank...\n",
      "Refseq species hash mappings already created, traversing online directories now...\n",
      "Invalid species name, cannot be found in Refseq alternate directory.\n",
      "Genbank species hash mappings already created, traversing online directories now...\n",
      "Alternative Genbank database route failed.\n",
      "Invalid species name or species does not have a protein annotation in either Genbank or Refseq.\n",
      "--------------------\n",
      "Warning: not all species had their protein annotations downloaded.\n",
      "Try manually downloading the necessary files and adding the species names and protein annotation file names to\n",
      "the absense_accession_codes_infile.txt file (or something else if it was changed).\n",
      "When satisfied, proceed with the evolutionary distance computation.\n"
     ]
    }
   ],
   "source": [
    "grab_assemblies(list_of_species = \n",
    "                ['Saccharomyces cerevisiae',\n",
    "                 'Saccharomyces mikatae',\n",
    "                 'Naumovozyma castellii',\n",
    "                 'Kluyveromyces lactis',\n",
    "                 'Aspergillus nidulans',\n",
    "                 'Saccharomyces kudriavzevii',\n",
    "                 'Yarrowia lipolytica',\n",
    "                 'Schizosaccharomyces pombe',\n",
    "                 'Eremothecium gossypii',\n",
    "                 'Saccharomyces paradoxus',\n",
    "                 'Saccharomyces bayanus'],\n",
    "                 override_dict = \n",
    "                 {'Saccharomyces mikatae' : ['Smik.faa' , 'target'],\n",
    "                  'Saccharomyces bayanus' : ['Sbay.faa', 'target']})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------------------\n",
      "Eukaryota ODB10 BUSCO genes already downloaded and successfully opened.\n",
      "-------------------\n",
      "Running HMMER calcuations now...\n",
      "-------------------\n",
      "Iterating results and grabbing orthologs now with E-value of 0.01...\n",
      "-------------------\n",
      "Creating files for MUSCLE now...\n",
      "-------------------\n",
      "Running MUSCLE now...\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_009533.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_009616.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:01    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_009729.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_009802.3 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_010014.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_010490.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_010584.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_010761.3 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_010777.3 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_011617.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_011727.3 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_011732.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_011876.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_011925.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_012230.3 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_012273.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_012355.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_012547.2 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_013273.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_013564.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:01    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:01    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:01    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:01    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:01    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:01    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_014337.3 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_014517.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_015081.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_015113.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_015176.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_015341.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_015361.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:01    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "NP_015365.1 11 seqs, lengths min 252, max 403, avg 340\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    24 MB(-3%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Align node       \n",
      "00:00:00    28 MB(-4%)  Iter   1  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Refine tree   \n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   2  100.00%  Root alignment\n",
      "00:00:00    28 MB(-4%)  Iter   3  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   4  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   5  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   6  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   7  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   8  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter   9  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  10  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  11  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  12  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  13  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  14  100.00%  Refine biparts\n",
      "00:00:00    28 MB(-4%)  Iter  15  100.00%  Refine biparts\n",
      "-------------------\n",
      "Running Protdist now...\n",
      "\u001b[2J\u001b[H\n",
      "\u001b[2J\u001b[H\n",
      "Protein distance algorithm, version 3.697\n",
      "\n",
      "Settings for this run:\n",
      "  P  Use JTT, PMB, PAM, Kimura, categories model?  Jones-Taylor-Thornton matrix\n",
      "  G  Gamma distribution of rates among positions?  No\n",
      "  C           One category of substitution rates?  Yes\n",
      "  W                    Use weights for positions?  No\n",
      "  M                   Analyze multiple data sets?  No\n",
      "  I                  Input sequences interleaved?  Yes\n",
      "  0                 Terminal type (IBM PC, ANSI)?  ANSI\n",
      "  1            Print out the data at start of run  No\n",
      "  2          Print indications of progress of run  Yes\n",
      "\n",
      "Are these settings correct? (type Y or the letter for one to change)\n",
      "\n",
      "Computing distances:\n",
      "  S_pombe      \n",
      "  K_lacti      .\n",
      "  N_caste      ..\n",
      "  E_gossy      ...\n",
      "  S_bayan      ....\n",
      "  S_kudri      .....\n",
      "  S_mikat      ......\n",
      "  S_cerev      .......\n",
      "  S_parad      ........\n",
      "  Y_lipol      .........\n",
      "  A_nidul      ..........\n",
      "\n",
      "Output written to file \"outfile\"\n",
      "\n",
      "Done.\n",
      "\n",
      "cp: warning: source file 'infile' specified more than once\n",
      "cp: warning: source file 'outfile' specified more than once\n",
      "-------------------\n",
      "Done! Results are in \"outfile\" in the ./protdist_infile_outfile_GCF_000146045.2_R64_protein directory. Thanks!\n"
     ]
    }
   ],
   "source": [
    "run_evo_dist_computation(evalue = 0.01, acc_file = 'absense_accession_codes_infile.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
